### PR TITLE
[PDI-19580]-Enable Use of OAUTH for PDI Mail Steps and Entries that use IMAP/SMTP

### PR DESCRIPTION
**Pull request by Mandakini111**
[PDI-19580]-Enable Use of OAUTH for PDI Mail Steps and Entries that use IMAP/SMTP

[PDI-19580]: https://hv-eng.atlassian.net/browse/PDI-19580?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by buildguy** [comment]
### :x: Build failed in 42m 45s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
plugins/mail-job/impl,plugins/mail/impl,plugins/mail/ui
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 2 test(s) failed:

<details><summary><b>org.pentaho.di.trans.steps.mail.MailTest.testSendMailWithPlainPassword</b> (click to expand)</summary><p>

```

Argument(s) are different! Wanted:
mail.sendMail(
    ["smtp.gmail.com", "sender@mail.com", "qwerty", "destination@mail.com", "Leonardo", "sender@mail.com", "Test Transform Auth Email", "This is a test email", "587"],
    "smtp.gmail.com",
    -1,
    "Leonardo",
    null,
    "destination@mail.com",
    null,
    null,
    null,
    null,
    "sender@mail.com",
    "qwerty",
    null,
    null,
    null
);
-> at org.pentaho.di.trans.steps.mail.Mail.sendMail(Mail.java:572)
Actual invocations have different arguments at positions [10, 11]:
mail.init(
    Mock for MailMeta, hashCode: 1386090898,
    Mock for MailData, hashCode: 847552559
);
-> at org.pentaho.di.trans.steps.mail.MailTest.testSendMailWithPassword(MailTest.java:212)
mail.getStepMeta(
    
);
-> at org.pentaho.di.trans.step.BaseStep.init(BaseStep.java:657)
mail.getStepMeta(
    
);
-> at org.pentaho.di.trans.step.BaseStep.init(BaseStep.java:659)
mail.setVariable(
    "Internal.Step.Unique.Number",
    "0"
);
-> at org.pentaho.di.trans.step.BaseStep.init(BaseStep.java:664)
mail.setVariable(
    "Internal.Step.Unique.Count",
    "1"
);
-> at org.pentaho.di.trans.step.BaseStep.init(BaseStep.java:665)
mail.setVariable(
    "Internal.Step.CopyNr",
    "0"
);
-> at org.pentaho.di.trans.step.BaseStep.init(BaseStep.java:666)
mail.setParentVariableSpace(
    org.pentaho.di.core.variables.Variables@86a4b9b
);
-> at org.pentaho.di.trans.steps.mail.MailTest.testSendMailWithPassword(MailTest.java:213)
mail.setInputRowMeta(
    [smtp_host String], [smtp_auth_user String], [smtp_auth_password String], [email_recipient String], [email_sender_name String], [email_sender_address String], [email_subject String], [email_body String], [smtp_port String]
);
-> at org.pentaho.di.trans.steps.mail.MailTest.testSendMailWithPassword(MailTest.java:234)
mail.setInputRowSets(
    [Mock for RowSet, hashCode: 903182919]
);
-> at org.pentaho.di.trans.steps.mail.MailTest.testSendMailWithPassword(MailTest.java:245)
mail.processRow(
    Mock for MailMeta, hashCode: 1555283355,
    org.pentaho.di.trans.steps.mail.MailData@3fe9caee
);
-> at org.pentaho.di.trans.steps.mail.MailTest.testSendMailWithPassword(MailTest.java:258)
mail.getRow(
    
);
-> at org.pentaho.di.trans.steps.mail.Mail.processRow(Mail.java:104)
mail.getRowHandler(
    
);
-> at org.pentaho.di.trans.step.BaseStep.getRow(BaseStep.java:1816)
mail.waitUntilTransformationIsStarted(
    
);
-> at org.pentaho.di.trans.step.BaseStep.handleGetRow(BaseStep.java:1842)
mail.openRemoteInputStepSocketsOnce(
    
);
-> at org.pentaho.di.trans.step.BaseStep.handleGetRow(BaseStep.java:1846)
mail.isUsingThreadPriorityManagment(
    
);
-> at org.pentaho.di.trans.step.BaseStep.handleGetRow(BaseStep.java:1884)
mail.isStopped(
    
);
-> at org.pentaho.di.trans.step.BaseStep.handleGetRow(BaseStep.java:1905)
mail.incrementLinesRead(
    
);
-> at org.pentaho.di.trans.step.BaseStep.handleGetRow(BaseStep.java:1912)
mail.getInputRowMeta(
    
);
-> at org.pentaho.di.trans.steps.mail.Mail.processRow(Mail.java:115)
mail.validateMetaFields(
    Mock for MailMeta, hashCode: 1555283355
);
-> at org.pentaho.di.trans.steps.mail.Mail.processRow(Mail.java:117)
mail.validateZipFiles(
    Mock for MailMeta, hashCode: 1555283355
);
-> at org.pentaho.di.trans.steps.mail.Mail.validateMetaFields(Mail.java:503)
mail.environmentSubstitute(
    null
);
-> at org.pentaho.di.trans.steps.mail.Mail.cacheFieldsPosition(Mail.java:463)
mail.environmentSubstitute(
    null
);
-> at org.pentaho.di.trans.steps.mail.Mail.cacheFieldsPosition(Mail.java:469)
mail.processAttachedFiles(
    
);
-> at org.pentaho.di.trans.steps.mail.Mail.cacheFieldsPosition(Mail.java:472)
mail.environmentSubstitute(
    null
);
-> at org.pentaho.di.trans.steps.mail.Mail.processAttachedFiles(Mail.java:535)
mail.environmentSubstitute(
    null
);
-> at org.pentaho.di.trans.steps.mail.Mail.processAttachedFiles(Mail.java:536)
mail.sendMail(
    ["smtp.gmail.com", "sender@mail.com", "qwerty", "destination@mail.com", "Leonardo", "sender@mail.com", "Test Transform Auth Email", "This is a test email", "587"],
    "smtp.gmail.com",
    -1,
    "Leonardo",
    null,
    "destination@mail.com",
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null
);
-> at org.pentaho.di.trans.steps.mail.Mail.processRow(Mail.java:193)
mail.getInputRowMeta(
    
);
-> at org.pentaho.di.trans.steps.mail.Mail.processRow(Mail.java:198)
mail.putRow(
    [smtp_host String], [smtp_auth_user String], [smtp_auth_password String], [email_recipient String], [email_sender_name String], [email_sender_address String], [email_subject String], [email_body String], [smtp_port String],
    ["smtp.gmail.com", "sender@mail.com", "qwerty", "destination@mail.com", "Leonardo", "sender@mail.com", "Test Transform Auth Email", "This is a test email", "587"]
);
-> at org.pentaho.di.trans.steps.mail.Mail.processRow(Mail.java:198)
mail.getRowHandler(
    
);
-> at org.pentaho.di.trans.step.BaseStep.putRow(BaseStep.java:1259)
mail.incrementLinesWritten(
    
);
-> at org.pentaho.di.trans.step.BaseStep.handlePutRow(BaseStep.java:1321)

```

</p></details>

<details><summary><b>org.pentaho.di.trans.steps.mail.MailTest.testSendMailWithEncryptedPassword</b> (click to expand)</summary><p>

```

Argument(s) are different! Wanted:
mail.sendMail(
    ["smtp.gmail.com", "sender@mail.com", "Encrypted 2be98afc86aa7f2e4cb79bf67db80bbc3", "destination@mail.com", "Leonardo", "sender@mail.com", "Test Transform Auth Email", "This is a test email", "587"],
    "smtp.gmail.com",
    -1,
    "Leonardo",
    null,
    "destination@mail.com",
    null,
    null,
    null,
    null,
    "sender@mail.com",
    "qwerty",
    null,
    null,
    null
);
-> at org.pentaho.di.trans.steps.mail.Mail.sendMail(Mail.java:572)
Actual invocations have different arguments at positions [10, 11]:
mail.init(
    Mock for MailMeta, hashCode: 488220733,
    Mock for MailData, hashCode: 1521530739
);
-> at org.pentaho.di.trans.steps.mail.MailTest.testSendMailWithPassword(MailTest.java:212)
mail.getStepMeta(
    
);
-> at org.pentaho.di.trans.step.BaseStep.init(BaseStep.java:657)
mail.getStepMeta(
    
);
-> at org.pentaho.di.trans.step.BaseStep.init(BaseStep.java:659)
mail.setVariable(
    "Internal.Step.Unique.Number",
    "0"
);
-> at org.pentaho.di.trans.step.BaseStep.init(BaseStep.java:664)
mail.setVariable(
    "Internal.Step.Unique.Count",
    "1"
);
-> at org.pentaho.di.trans.step.BaseStep.init(BaseStep.java:665)
mail.setVariable(
    "Internal.Step.CopyNr",
    "0"
);
-> at org.pentaho.di.trans.step.BaseStep.init(BaseStep.java:666)
mail.setParentVariableSpace(
    org.pentaho.di.core.variables.Variables@5bc971ef
);
-> at org.pentaho.di.trans.steps.mail.MailTest.testSendMailWithPassword(MailTest.java:213)
mail.setInputRowMeta(
    [smtp_host String], [smtp_auth_user String], [smtp_auth_password String], [email_recipient String], [email_sender_name String], [email_sender_address String], [email_subject String], [email_body String], [smtp_port String]
);
-> at org.pentaho.di.trans.steps.mail.MailTest.testSendMailWithPassword(MailTest.java:234)
mail.setInputRowSets(
    [Mock for RowSet, hashCode: 1262117763]
);
-> at org.pentaho.di.trans.steps.mail.MailTest.testSendMailWithPassword(MailTest.java:245)
mail.processRow(
    Mock for MailMeta, hashCode: 679567512,
    org.pentaho.di.trans.steps.mail.MailData@16ec1438
);
-> at org.pentaho.di.trans.steps.mail.MailTest.testSendMailWithPassword(MailTest.java:258)
mail.getRow(
    
);
-> at org.pentaho.di.trans.steps.mail.Mail.processRow(Mail.java:104)
mail.getRowHandler(
    
);
-> at org.pentaho.di.trans.step.BaseStep.getRow(BaseStep.java:1816)
mail.waitUntilTransformationIsStarted(
    
);
-> at org.pentaho.di.trans.step.BaseStep.handleGetRow(BaseStep.java:1842)
mail.openRemoteInputStepSocketsOnce(
    
);
-> at org.pentaho.di.trans.step.BaseStep.handleGetRow(BaseStep.java:1846)
mail.isUsingThreadPriorityManagment(
    
);
-> at org.pentaho.di.trans.step.BaseStep.handleGetRow(BaseStep.java:1884)
mail.isStopped(
    
);
-> at org.pentaho.di.trans.step.BaseStep.handleGetRow(BaseStep.java:1905)
mail.incrementLinesRead(
    
);
-> at org.pentaho.di.trans.step.BaseStep.handleGetRow(BaseStep.java:1912)
mail.getInputRowMeta(
    
);
-> at org.pentaho.di.trans.steps.mail.Mail.processRow(Mail.java:115)
mail.validateMetaFields(
    Mock for MailMeta, hashCode: 679567512
);
-> at org.pentaho.di.trans.steps.mail.Mail.processRow(Mail.java:117)
mail.validateZipFiles(
    Mock for MailMeta, hashCode: 679567512
);
-> at org.pentaho.di.trans.steps.mail.Mail.validateMetaFields(Mail.java:503)
mail.environmentSubstitute(
    null
);
-> at org.pentaho.di.trans.steps.mail.Mail.cacheFieldsPosition(Mail.java:463)
mail.environmentSubstitute(
    null
);
-> at org.pentaho.di.trans.steps.mail.Mail.cacheFieldsPosition(Mail.java:469)
mail.processAttachedFiles(
    
);
-> at org.pentaho.di.trans.steps.mail.Mail.cacheFieldsPosition(Mail.java:472)
mail.environmentSubstitute(
    null
);
-> at org.pentaho.di.trans.steps.mail.Mail.processAttachedFiles(Mail.java:535)
mail.environmentSubstitute(
    null
);
-> at org.pentaho.di.trans.steps.mail.Mail.processAttachedFiles(Mail.java:536)
mail.sendMail(
    ["smtp.gmail.com", "sender@mail.com", "Encrypted 2be98afc86aa7f2e4cb79bf67db80bbc3", "destination@mail.com", "Leonardo", "sender@mail.com", "Test Transform Auth Email", "This is a test email", "587"],
    "smtp.gmail.com",
    -1,
    "Leonardo",
    null,
    "destination@mail.com",
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null
);
-> at org.pentaho.di.trans.steps.mail.Mail.processRow(Mail.java:193)
mail.getInputRowMeta(
    
);
-> at org.pentaho.di.trans.steps.mail.Mail.processRow(Mail.java:198)
mail.putRow(
    [smtp_host String], [smtp_auth_user String], [smtp_auth_password String], [email_recipient String], [email_sender_name String], [email_sender_address String], [email_subject String], [email_body String], [smtp_port String],
    ["smtp.gmail.com", "sender@mail.com", "Encrypted 2be98afc86aa7f2e4cb79bf67db80bbc3", "destination@mail.com", "Leonardo", "sender@mail.com", "Test Transform Auth Email", "This is a test email", "587"]
);
-> at org.pentaho.di.trans.steps.mail.Mail.processRow(Mail.java:198)
mail.getRowHandler(
    
);
-> at org.pentaho.di.trans.step.BaseStep.putRow(BaseStep.java:1259)
mail.incrementLinesWritten(
    
);
-> at org.pentaho.di.trans.step.BaseStep.handlePutRow(BaseStep.java:1321)

```

</p></details>

  
Tests run: **33**, Failures: **2**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9717/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
[comment]: <> (FrogbotReviewComment)

<div align='center'>

[👍 Frogbot scanned this pull request and did not find any new security issues.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 18m 22s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
plugins/mail-job/impl,plugins/mail/impl,plugins/mail/ui
```
---
:ok_hand: All tests passed!
 
Tests run: **33**, Failures: **0**, Skipped: **2** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9717/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: plugins/mail/impl/src/main/java/org/pentaho/di/trans/steps/mail/MailMeta.java [hunk 1]
```diff
@@ -285,11 +324,21 @@ public String getXML() throws KettleException {
...
     retval.append( "      " ).append( XMLHandler.addTagValue( "use_auth", this.usingAuthentication ) );
+    retval.append( "      " ).append( XMLHandler.addTagValue( "use_grantType", grant_type) );
     retval.append( "      " ).append( XMLHandler.addTagValue( "use_secure_auth", this.usingSecureAuthentication ) );
     retval.append( "      " ).append( XMLHandler.addTagValue( "auth_user", this.authenticationUser ) );
     retval.append( "      " ).append(
       XMLHandler.addTagValue( "auth_password", Encr
         .encryptPasswordIfNotUsingVariables( this.authenticationPassword ) ) );
+    retval.append( "      " ).append( XMLHandler.addTagValue( "auth_clientId", clientId ) );
+    retval.append( "      " ).append(
+            XMLHandler
```
Comment (peterrinehart) [inline plugins/mail/impl/src/main/java/org/pentaho/di/trans/steps/mail/MailMeta.java:335]: Why is the extra whitespace here?

**Review by peterrinehart — CHANGES_REQUESTED** [review]

**Comment by buildguy** [comment]
[comment]: <> (FrogbotReviewComment)

<div align='center'>

[👍 Frogbot scanned this pull request and did not find any new security issues.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 21m 7s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl \
plugins/mail-job/impl,plugins/mail/impl,plugins/mail/ui
```
---
:ok_hand: All tests passed!
 
Tests run: **33**, Failures: **0**, Skipped: **2** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9717/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Pull request merged**